### ISSUE TITLE
This KCL locks code mirror completely

### ISSUE THREAD

**Issue by lf94** [KittyCAD/modeling-app#4327]
```
// L-System KCL
// Zoo Corporation ⓒ 2024


// L-system 
// Note: it was most concise to encode productions directly in axioms.
// Change them as you need.

deg = pi()*2 / 360

fn setSketch = (state, _) => {
  return {
    depthMax: state.depthMax,
    depth: state.depth + 1,
    currentLength: state.currentLength,
    factor: state.factor,
    currentAngle: state.currentAngle,
    angle: state.angle,
    _: _
  }
}

fn setDepth = (state, _) => {
  return {
    depthMax: state.depthMax,
    depth: _,
    currentLength: state.currentLength,
    factor: state.factor,
    currentAngle: state.currentAngle,
    angle: state.angle,
    _:  state._
  }
}

fn setAngle = (state, _) => {
  return {
    depthMax: state.depthMax,
    depth: state.depth,
    currentLength: state.currentLength,
    factor: state.factor,
    currentAngle: _,
    angle: state.angle,
    _:  state._
  }
}

fn setLength = (state, _) => {
  return {
    depthMax: state.depthMax,
    depth: state.depth,
    currentLength: _,
    factor: state.factor,
    currentAngle: state.currentAngle,
    angle: state.angle,
    _:  state._
  }
}

fn Gt2 = (state) => { return setLength(state, state.currentLength * state.factor) }
fn Lt2 = (state) => { return setLength(state, state.currentLength / state.factor) }
fn Add = (state) => { return setAngle(state, rem(int(state.currentAngle - state.angle), 360)) }
fn Sub = (state) => { return setAngle(state, rem(int(state.currentAngle + state.angle), 360)) }

// Only necessary to get around recursion limitations...
fn X = (state, X) => {
  return if state.depth < state.depthMax {
    stateNext = state |> setDepth(%, state.depth + 1)

    // Produce
    // Note:if you need [ and ], just save state to a variable.
    stateNext
      |> F(%, F) |> Sub(%) |> F(%, F)
      |> Add(%) |> Add(%)
      |> F(%, F) |> Sub(%) |> F(%, F)
      |> setDepth(%, stateNext.depth - 1)

  } else {
    // Pass onto the next instruction
    state |> setSketch(%, angledLine({ angle: state.currentAngle, length: state.currentLength }, state._))
  }
}

// Only necessary to get around recursion limitations...
fn F = (state, F) => {
  
    // Pass onto the next instruction
    state |> setSketch(%, angledLine({ angle: state.currentAngle, length: state.currentLength }, state._))
  
}

fn LSystem = (args, axioms) => {
  return axioms({
    depthMax: args.iterations,
    depth: 0,
    currentLength: 1.0,
    factor: args.factor,
    currentAngle: 0,
    angle: args.angle,
    _: startSketchAt([0, 0]),
  })
}

LSystem({
  iterations: 2,
  factor: 1.36,
  angle: 60,
}, (_) => {
  result = _ |> F(%, F) |> Add(%) |> Add(%) |> F(%, F) |> Add(%) |> Add(%) |> F(%, F) 
  return result._
})
```

**Comment by jtran** [comment]
It's quite frustrating that any problem causes CodeMirror to lock up, meaning the user can't get out of the problematic state. #4240 is another example.

**Comment by lf94** [comment]
Holy f***ing hell. The problem here is *the parser is reporting an out-of-bounds source range because it's interpreting text as bytes and not characters. The copyright symbol is technically 3 bytes, but 1 character, making the source range always off by 2.* This has to be fixed on the parser side.

**Comment by jtran** [comment]
We think this is related to the transcoding of UTF-16 JS `string` to the UTF-8 Rust `&str`. See [wasm_bindgen docs](https://rustwasm.github.io/docs/wasm-bindgen/reference/types/str.html).

**Comment by lf94** [comment]
How the hell do we solve this one lol. Still around today.

**Comment by jtran** [comment]
The Rust side is computing UTF-8 byte offsets. JS is using this and treating it as code unit offsets. For most text, you can think of these as character offsets. But UTF-16 has [surrogate pairs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String#utf-16_characters_unicode_code_points_and_grapheme_clusters), so these are not quite characters. An emoji, for example, takes two indices in terms of `str.slice()` or string indexing. You need to be careful not to slice an emoji in half.

We should be able to do the mapping on either side. Since we need to make the WASM binding on the Rust side and properly deserialize errors on the TS side, we basically already have a wrapper function on both sides of the boundary where we could put this.

The most straight-forward way to map a single offset is to take the slice of the original string up to the offset you care about, transcode it, and see how long the result is. In JS, you could use [TextDecoder](https://developer.mozilla.org/en-US/docs/Web/API/TextDecoder) to do this.

But for us, we have multiple offsets that we want to compute. Two for each source range, and we may have many source ranges, one for each warning or error. We don't allow multiple errors yet, but we will soon.

Say we have _k_ offsets into a string of length _n_. If you used the above approach and naively wrapped it in a function, calling it _k_ times, it would be O(_kn_), which is not great. We'd want to keep a running count and compute all _k_ sequentially in byte-offset order, accumulating on the previous one, leading to O(_n_).

But before we implement anything, is there a completely different approach we could be using? I know we've explored them, but this is another reason to (re-)consider them.

**Comment by lf94** [comment]
https://simonsapin.github.io/wtf-8/

**Comment by adamchalmers** [comment]
I have a shorter repro of this bug.

```
亞當 = 3
x = 亞當 * "a"
startSketchOn(XY)
|> startProfile(at = [0, 0])
```

<img width="279" alt="Image" src="https://github.com/user-attachments/assets/904e547d-d153-4b14-9087-d85a1c4ecd25" />

Note that the error source range (red squiggly underline) is inaccurately in the middle of "startSketch", when it should be on the `* "a"`. It's calculating a source range which is too far to the right. If the program didn't have characters that far to the right, then the error doesn't get highlighted at all (there's no code to highlight there). But it _does_ show up in a toast and console.error:

<img width="546" alt="Image" src="https://github.com/user-attachments/assets/00e53c22-c070-43fa-bb3a-62b575ad6bf8" />

**Comment by adamchalmers** [comment]
Interesting! @jtran @lf94 this is not just a UTF-8 vs. UTF-16 issue. 

```rust
fn main() {
    let msg = "亞當 = 3
x = 亞當 * \"a\"
startSketchOn(XY)
|> startProfile(at = [0, 0])\n";

    let num_chars = msg.len();
    let utf8_len = msg.chars().count();
    let utf16_len = msg.encode_utf16().count();
    assert_eq!(num_chars, 74);
    assert_eq!(utf8_len, 66);
    assert_eq!(utf16_len, 66);
}
```

[(playground link)](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=6f56fc5731e0890564fb4cce3fa689f0)

UTF-8 and UTF-16 agree that the program I posted above is 66 bytes. So that's not the problem. I think the issue is that the SourceRange is computed as byte offsets in the UTF-8 bytes, but is being displayed as character offsets in the string. 

Now, if we changed the message, say by adding some weird emojis, then we do have a UTF-8 vs 16 issue.

So, two changes:

- Frontend and Rust need a contract about what a source range is:
  - a character offset
  - a UTF-8 byte offset
  - a UTF-16 byte offset
- Frontend should display diagnostics based on that contract

**Comment by jtran** [comment]
Makes sense. `msg.len()` is the number of bytes, not the number of code points.
